notify/fuse: first working implementation

rjeczalik opened the pull request:
This PR adds first working implementation of FUSE-based notifier for a mount subsystem.

Deps: ~#10588 #10590~

There's going to be another PR with cmd/loopfuse - a tool I used to test current implementation. Since rsync-based syncer is not yet integrated, I did not write any integration tests just to not have to rework them after the rsync is in. The integration tests we're going to write separately after gathering feedback from more wide internal testing.

rjeczalik committed:
notify/fuse: index-backed filesystem implementation

rjeczalik commented:
I think something change with github and I'm not longer able to create forward PRs that depend on others - trying to create a PR from koding:loopfuse branch fails with:

```
< HTTP 422
{"message":"Validation Failed","errors":[{"resource":"PullRequest","code":"custom","message":"A pull request already exists for koding:loopfuse."}],"documentation_url":"https://developer.github.com/v3/pulls/#create-a-pull-request"}
Error creating pull request: Unprocessable Entity (HTTP 422)
A pull request already exists for koding:loopfuse.
```

The tool waits at this branch:

https://github.com/koding/koding/tree/loopfuse/go/src/koding/klient/machine/mount/notify/fuse/cmd/loopfuse

I'm going to send a PR after this one is merged.

ppknap commented:
@rjeczalik could you rebase when #10588 is merged?

rjeczalik commented:
@ppknap Rebased.

ppknap commented on go/src/koding/klient/machine/mount/notify/fuse/filesystem.go:
```diff
@@ -0,0 +1,624 @@
+		return nil
+	}
+
+	return fs.yield(ctx, path, index.ChangeMetaLocal|index.ChangeMetaRemove)
```
we don't need to wait here for update

rjeczalik replied:
Done.

ppknap commented on go/src/koding/klient/machine/mount/notify/fuse/filesystem.go:
```diff
@@ -0,0 +1,624 @@
+
+	updateSize(f, nd)
+
+	return fs.yield(ctx, f.Name(), index.ChangeMetaLocal|index.ChangeMetaUpdate)
```
no need to wait here

rjeczalik replied:
Done.

ppknap commented on go/src/koding/klient/machine/mount/notify/fuse/fuse.go:
```diff
@@ -0,0 +1,481 @@
+
+	op.Entry.Attributes = fs.newAttr(op.Mode)
+
+	return fs.yield(ctx, path, index.ChangeMetaAdd|index.ChangeMetaLocal)
```
do we need to wait here?

rjeczalik replied:
Done.

ppknap commented on go/src/koding/klient/machine/mount/notify/fuse/fuse.go:
```diff
@@ -0,0 +1,481 @@
+
+	op.Entry.Attributes = fs.newAttr(op.Mode)
+
+	return fs.yield(ctx, path, index.ChangeMetaAdd|index.ChangeMetaLocal)
```
no need to wait here.

rjeczalik replied:
Done.

ppknap commented on go/src/koding/klient/machine/mount/notify/fuse/fuse.go:
```diff
@@ -0,0 +1,481 @@
+	fs.Remote.PromiseAdd(newPath, entry)
+
+	return nonil(
+		fs.yield(ctx, oldPath, index.ChangeMetaLocal|index.ChangeMetaRemove),
```
no need to wait here
```go
fs.Cache.Commit(index.NewChange(oldpath, index.ChangeMetaRemove))
fs.Cache.Commit(index.NewChange(newpath, index.ChangeMetaAdd))

return nil
```

rjeczalik replied:
Done.

ppknap commented on go/src/koding/klient/machine/mount/notify/fuse/fuse.go:
```diff
@@ -0,0 +1,481 @@
+		return err
+	}
+
+	return fs.yield(ctx, path, index.ChangeMetaLocal|index.ChangeMetaRemove)
```
no wait

rjeczalik replied:
Done.

ppknap commented on go/src/koding/klient/machine/mount/notify/fuse/fuse.go:
```diff
@@ -0,0 +1,481 @@
+
+	updateSize(f, nd)
+
+	return fs.yield(ctx, f.Name(), index.ChangeMetaLocal|index.ChangeMetaUpdate)
```
no need to wait

rjeczalik replied:
Done.

ppknap reviewed (commented):
FUSE doesn't have to wait for all syncs since it has some files locally available just after the call

cihangir commented on go/src/koding/klient/machine/mount/notify/fuse/export_test.go:
```diff
@@ -0,0 +1,4 @@
+package fuse
+
+// TrimRightNull exports trimRightNull for testing purposes.
```
👍

rjeczalik committed:
notify/fuse: filesystem improvements

    - fix entry merging / copying
    - display . in directory listing
    - add IOSize
    - copy metadata on rename
    - use only 1 fd for mkfile
    - fix permission overwrite

rjeczalik commented on go/src/koding/klient/machine/mount/notify/fuse/filesystem.go:
```diff
@@ -0,0 +1,624 @@
+		return 0, 0, err
+	}
+
+	if f, err = os.OpenFile(absPath, os.O_RDWR, 0644); err != nil {
```
Done, but different way round - removed the second open.

rjeczalik committed:
notify/fuse: do not wait on cache writes

rjeczalik commented:
@ppknap @cihangir All addressed, ptal.

rjeczalik committed:
klient/fs: add Statfs

cihangir commented:
i haven't checked this PR since it is bigger than our limits, is there any way to decrease the size and add unit tests?

rjeczalik commented:
@cihangir I can rework it into PRs, each than contain only one method implementation for fuse.FileSystem. It was not possible to do that from the begining, since I needed working fuse implementation to ensure the code od correct. As it was not possible to prove it with unittests, I've created a tool - loopfuse - in a separate PR. I will add also unittests for the methods as well.

ppknap commented on go/src/koding/klient/fs/statfs_linux.go:
```diff
@@ -0,0 +1,20 @@
+	di := &DiskInfo{
+		BlockSize:   uint32(stfs.Bsize),
+		BlocksTotal: stfs.Blocks,
+		BlocksFree:  stfs.Bfree,
```
Is there a reason why `IOSize` is missing here? I used it on my linux machine.

rjeczalik replied:
https://github.com/golang/go/blob/master/src/syscall/ztypes_linux_amd64.go#L101-L117

How?

ppknap commented:
>  I will add also unittests for the methods as well.

Could it be done in separate PR after this is merged? @cihangir this PR may be hard to split into smaller pieces. If this help: `filesystem.go` is a priv interface for `fuse.go` impl (which implements `jacobsa/go-fuse`). Rest of the files are not directly related. This was my approach to review this (quite large) repo.

It would be very nice to have this merged as soon as possible.

cihangir commented:
It seems you want this PR in so bad @ppknap :)

> Could it be done in separate PR after this is merged?

historically we dont do that most of the time, something _more_ important comes up and we postpone it to a _later_ time.

I would like to see PRs with their unit tests next to them, so my vote is 👎 for merging. But if you both guys say OK to merge you can go ahead and merge. I have reviewed PRs three times bigger than this so the size doesnt matter i just dont want to break the rules we agreed upon.

rjeczalik commented:
@cihangir @ppknap We have two options:

1) have a devel branch for mounts, develop there until we reach a beta version that can be deployed to dev channel and be tested by QA; then later rework it to be production-ready code and send nice PRs
2) merge it now and rework later on master

I always opt for 1) as prevents merging code to master and then forgetting to rework it, which happens way too many times.

I will rework* FUSE after I'm done with 1) kloud 2) kite fixes ("unknown error occurred") 3) kdio. More details in an internal mail that I will send later.

* scout's word

ppknap commented:
> It seems you want this PR in so bad @ppknap :)

Hah, you discovered my evil plan of merging this. I'm bored of being HTML developer :)

> have a devel branch for mounts,

I spent about one hour fixing >100 merge conflicts in >15 commits. Trying to avoid this.

rjeczalik commented:
> I spent about one hour fixing >100 merge conflicts in >15 commits. Trying to avoid this.

Becase we're not developing against devel branch.

You create a devel branch, branch off your fork branch, work on a PR, after it's ready and won't be touched again you push it to devel. Then rebase your another feature branch against devel. I did not have excessive merge conflicts following that flow. Unless I happen to edit the same file that my fellow developer.

After everything is ready we slice the devel branch in PRs and send in order as gh PRs against master.

cihangir commented:
https://github.com/koding/koding/pull/10591#issuecomment-282761694

If you say we will make it better, we could merge it to master as well.

ppknap commented:
@rjeczalik before we merge, some of `Commit` calls get an absolute path of the change and some get relative. Could you fix this and make all of them relative to cache directory?

rjeczalik committed:
fuse: fix rel/abs path handling

rjeczalik commented:
@ppknap Done in 1a072415f5be8309b3a55989e6990ba60e3e5fc3.

ppknap commented:
@cihangir @rjeczalik what about merging this? I will tests these changes anyway.

![image](https://thehftguy.files.wordpress.com/2017/02/ship-crashing-in-the-sea.gif?w=809)
